base/inference.jl

```diff
@@ -2335,14 +2335,32 @@ end
 
 #### post-inference optimizations ####
 
-function inline_as_constant(val::ANY, argexprs, linfo::LambdaInfo)
+immutable InvokeData
+    mt::MethodTable
+    entry::TypeMapEntry
+    types0
+    fexpr
+    texpr
+end
+
+function inline_as_constant(val::ANY, argexprs, linfo::LambdaInfo,
+                            invoke_data::ANY)
+    invoke_fexpr, invoke_texpr = if invoke_data === nothing
```

Comment:
This order of statements seems harder to read to me than just the straightforward conditionals and assignments